features/feature_case/asm/asm_mma.cu

```diff
@@ -0,0 +1,387 @@
+    }
+
+    b[i] = B[B_OFFSET + OFFSET((C_THREAD_COL * 2) + (i & 0x1) + r_off, C_THREAD_ROW, N)];
+  }
+
+  for (int i = 0; i < 4; i++) {
+    if (i < 2) {
+      c[i] = C[CD_OFFSET + OFFSET(C_THREAD_ROW, (C_THREAD_COL * 2) + (i & 0x1), N)];
+    } else {
+      c[i] = C[CD_OFFSET + OFFSET(C_THREAD_ROW + 8, (C_THREAD_COL * 2) + (i & 0x1), N)];
+    }
+  }
+
+  asm("mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 "
+      " { %0, %1, %2, %3 }, "
+      " { %4, %5, %6, %7 }, "
+      " { %8, %9 }, "
+      " { %10, %11, %12, %13 };"
+      : "+f"(d[0]), "+f"(d[1]), "+f"(d[2]), "+f"(d[3])
+      : "r"(*(reinterpret_cast<int *>(&ra[0]))),
+        "r"(*(reinterpret_cast<int *>(&ra[1]))),
+        "r"(*(reinterpret_cast<int *>(&ra[2]))),
+        "r"(*(reinterpret_cast<int *>(&ra[3]))),
+        "r"(*(reinterpret_cast<int *>(&rb[0]))),
+        "r"(*(reinterpret_cast<int *>(&rb[1]))),
+        "f"(c[0]), "f"(c[1]), "f"(c[2]), "f"(c[3]));
+
+  for (int i = 0; i < 4; i++) {
+    if (i < 2) {
+      D[CD_OFFSET + OFFSET(C_THREAD_ROW, (C_THREAD_COL * 2) + (i & 0x1), N)] = d[i];
+    } else {
+      D[CD_OFFSET + OFFSET(C_THREAD_ROW + 8, (C_THREAD_COL * 2) + (i & 0x1), N)] = d[i];
+    }
+  }
+}
+
+__global__ void mma_kernel_m16n8k16_s8_s32(int8_t *A, int8_t *B, int *C, int *D, int M, int N, int K) {
+  const int WARP_ID = threadIdx.x / WARP_SIZE;
+  const int LANE_ID = threadIdx.x % WARP_SIZE;
+
+  const int C_THREAD_ROW = LANE_ID / 4;
+  const int C_THREAD_COL = LANE_ID % 4;
+
+  int A_OFFSET = WARP_ID * M * K;
+  int B_OFFSET = WARP_ID * K * N;
+  int CD_OFFSET = WARP_ID * M * N;
+
+  int a[2] = {0};
+  int b = 0;
+  int c[4] = {0};
+  int d[4] = {0};
+
+  auto *ra = reinterpret_cast<int8_t *>(a);
+  auto *rb = reinterpret_cast<int8_t *>(&b);
+
+  for (int i = 0; i < 8; i++) {
+    int r_off = 0;
+    if (i >= 4) {
+      r_off = 8;
+    }
+    ra[i] = A[A_OFFSET + OFFSET(C_THREAD_ROW + r_off, (C_THREAD_COL * 4) + (i & 0x3), K)];
+  }
+
+  for (int i = 0; i < 4; i++) {
+    rb[i] = B[B_OFFSET + OFFSET((C_THREAD_COL * 4) + i, C_THREAD_ROW, N)];
+  }
+
+  for (int i = 0; i < 4; i++) {
+    if (i < 2)
+      c[i] = C[CD_OFFSET + OFFSET(C_THREAD_ROW, (C_THREAD_COL * 2) + (i & 0x1), N)];
+    else
+      c[i] = C[CD_OFFSET + OFFSET(C_THREAD_ROW + 8, (C_THREAD_COL * 2) + (i & 0x1), N)];
+  }
+
+  asm("mma.sync.aligned.m16n8k16.row.col.s32.s8.s8.s32 "
+      " { %0, %1, %2, %3 }, "
+      " { %4, %5 }, "
+      " { %6 }, "
+      " { %7, %8, %9, %10 };"
+      : "=r"(d[0]), "=r"(d[1]), "=r"(d[2]), "=r"(d[3])
+      : "r"(a[0]), "r"(a[1]),
+        "r"(b),
+        "r"(c[0]), "r"(c[1]), "r"(c[2]), "r"(c[3]));
+
+  for (int i = 0; i < 4; i++) {
+    if (i < 2)
+      D[CD_OFFSET + OFFSET(C_THREAD_ROW, (C_THREAD_COL * 2) + (i & 0x1), N)] = d[i];
+    else
+      D[CD_OFFSET + OFFSET(C_THREAD_ROW + 8, (C_THREAD_COL * 2) + (i & 0x1), N)] = d[i];
+  }
+}
+
+bool mma_m16n8k16_f16_f32() {
+  const int M = 16;
+  const int N = 8;
+  const int K = 16;
```

Comment:
Please also add other new test cases in which the values for `M, N, K` are integer multiples of `16, 8, 16`, like `32, 16, 32`, `32, 32, 32`.

Reply:
Done